Machine Learning Compute: Add -IncludeAllResources Paramter for Remove cmdlet

shutchings opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
Added a parameter to include all resources when removing an operationalization cluster.

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

shutchings commented:
@maddieclayton I've started this PR so that you can take a look at the code changes while I'm waiting for the [.NET SDK PR](https://github.com/Azure/azure-sdk-for-net/pull/3959) to be completed and the new package to be created. Once that's done I'll remove the local nuget package I added and make sure the build works.

maddieclayton commented on src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/RemoveAzureRmMlOpCluster.cs:
```diff
@@ -58,26 +58,47 @@ public class RemoveAzureRmMlOpCluster : MachineLearningComputeCmdletBase
+                var clusterToDelete = MachineLearningComputeManagementClient.OperationalizationClusters.Get(ResourceGroupName, Name);
+                var managedByResourceGroup = new ResourceIdentifier(clusterToDelete.ContainerRegistry.ResourceId).ResourceGroupName;
+
+                shouldProcessMessage += $" and supporting resource group {managedByResourceGroup}";
```
Is the supporting resource group the only additional resource that will be deleted when -IncludeAllResources is specified?

shutchings replied:
Yes it deletes that cluster and the supporting resource group only. The supporting resource group has a bunch of resources in it, I could change it to list them all out but it will include things like network security groups, nics, vms, disks, etc. It seemed too verbose to list all those resources individually so I think just warning that that resource group will be deleted should be enough.

maddieclayton replied:
I see, I agree that this message would make the most sense.  Maybe it would be useful to also add in the message that all resources in the supporting resource group will be deleted.

shutchings replied:
Good suggestion.

maddieclayton commented on src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Commands.MachineLearningCompute.csproj:
```diff
@@ -64,9 +64,12 @@
+    <Reference Include="Microsoft.Azure.Management.MachineLearningCompute, Version=0.3.0.0, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.MachineLearningCompute.0.3.0\lib\net452\Microsoft.Azure.Management.MachineLearningCompute.dll</HintPath>
     </Reference>
+    <Reference Include="Microsoft.Rest.ClientRuntime, Version=2.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL" />
```
Remove these three references: they are already imported by the common dependencies target.

maddieclayton commented on src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/packages.config:
```diff
@@ -1,4 +1,7 @@
 <packages>
-  <package id="Microsoft.Azure.Management.MachineLearningCompute" version="0.2.0" targetFramework="net452" />
+  <package id="Microsoft.Azure.Management.MachineLearningCompute" version="0.3.0" targetFramework="net452" />
+  <package id="Microsoft.Rest.ClientRuntime" version="2.3.10" targetFramework="net452" />
```
Same comment about removing the three duplicate references.

maddieclayton commented on src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute.Test/ScenarioTests/MLCTests.ps1:
```diff
@@ -302,4 +302,27 @@ function Test-Set
+	# Delete the cluster
+	Remove-AzureRmMlOpCluster -ResourceGroupName $resourceGroupName -Name $clusterName -IncludeAllResources
+
+    Assert-Throws ( Get-AzureRmResourceGroup -ResourceGroupName $managedByResourceGroupName )
```
Why does this throw? Shouldn't it complete the operation successfully and delete both the managed resource group and the cluster? And if it is supposed to throw, I would think you'd want another test with the command running successfully.

shutchings replied:
This throws because after a successful remove call with -IncludeAllResources the resource group I'm trying to get should not exist. So by expecting the throw I can verify that the -IncludeAllResources worked correctly.

maddieclayton replied:
Ah, that makes sense.  This is fine then.

maddieclayton commented:
@shutchings I'll take a look once the merge conflict has been resolved - please feel free to assign the PR back to me once this is done and you are ready for me to take a look.

shutchings commented:
@maddieclayton I fixed the merge conflict and removed the local package since I got the latest one published to nuget. From the looks of it it's failing static analysis and should regen the installer. Is that correct?

maddieclayton commented:
@shutchings The issue is that the parameter you added is plural (which is okay since it is a SwitchParameter).  Please copy the message (the issue with severity 1) from SignatureIssues.csv (found in the build) in [breakingchanges.csv](https://github.com/Azure/azure-powershell/blob/preview/tools/StaticAnalysis/Exceptions/BreakingChangeIssues.csv).  This should resolve the issue, please let me know if you have any more questions.

shutchings commented:
@maddieclayton I added in the exception but it is still failing. The other issues were severity 2 so I assume they aren't failing the build. Is there something else missing?

maddieclayton commented:
@shutchings No, the issue is still the plural parameter name.  Can you try following the format of the lines above Microsoft.Azure.Commands.Compute: 
````
"d:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Sql\Microsoft.Azure.Commands.Sql.dll","Microsoft.Azure.Commands.Sql.Database.Cmdlet.SetAzureSqlDatabase","Set-AzureRmSqlDatabase","0","2100","The parameter 'Tags' in cmdlet 'Set-AzureRmSqlDatabase' is no longer in the parameter set '__AllParameterSets'.","Add parameter 'Tags' back to the parameter set '__AllParameterSets'."
````
It looks like the format you used only works for compute for reason - if this does not fix the issue I'll look into it further.

shutchings commented:
@maddieclayton I think I have it in the right format now but it's still giving issues.

maddieclayton commented:
@shutchings Oh no, this is my fault.  It should go in tool/Exception/SignatureIssues.csv, not breakingchanges.  I apologize for the confusion, but that should work.  Don't use the new format that you switched to, just copy paste again into the csv file.

shutchings commented:
@maddieclayton Yep, that did it!

maddieclayton commented:
@shutchings Great! If you are ready for another review please assign the PR back to me.

maddieclayton commented on src/ResourceManager/MachineLearningCompute/ChangeLog.md:
```diff
@@ -18,10 +18,14 @@
+* Add IncludeAllResources parameter to Remove-AzureRmMlOpCluster cmdlet
+    - Using this switch parameter will remove all resources that were created with the cluster originally
+
+## Version 0.3.1
```
Don't modify this - just add your snippet after the snippets about ResourceGroup and Location completers.  We will automatically modify this file during release.

shutchings replied:
That makes sense. I looked at the version number in the .psd1 file and it shows 0.3.1 which is below the current release so that's why I thought the version numbers are off. But maybe I'm misunderstanding?

maddieclayton replied:
"Current release" actually means the upcoming release: it will be displayed as the current release notes in the next release.

maddieclayton commented on src/ResourceManager/MachineLearningCompute/Commands.MachineLearningCompute/Cmdlets/RemoveAzureRmMlOpCluster.cs:
```diff
@@ -58,26 +58,47 @@ public class RemoveAzureRmMlOpCluster : MachineLearningComputeCmdletBase
             HelpMessage = ResourceIdParameterHelpMessage)]
         public string ResourceId { get; set; }
 
+        [Parameter(ParameterSetName = ObjectParameterSet,
```
If the parameter is in all of the parameter sets you don't need a parameter attribute for each parameter set, all you need is:
```
[Parameter(Mandatory = false, HelpMessage = IncludeAllResourcesParameterHelpMessage)]
public SwitchParameter IncludeAllResources { get; set; }
````

shutchings replied:
Good to know!

maddieclayton commented:
on demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/320/https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/320/

shutchings commented:
@maddieclayton Is it possible to start a package job?

maddieclayton commented:
Yup, here is it:https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/posh-pack/42/.  Let me know when you are ready to merge.

maddieclayton commented:
@shutchings is this ready to merge? I wasn't sure if you were running testing on the package job or needed it for another reason.

shutchings commented:
@maddieclayton I found a typo in the parameter name in the help document while testing the package job. Must have been from when I regenerated the help and then made changes to the parameter name. After this build feel free to merge it.
